crates/agentgateway/src/parse/transform.rs

```diff
@@ -0,0 +1,159 @@
+
+use crate::*;
+
+pin_project! {
+	pub struct TransformedBody<D, E, F, T> {
+		#[pin]
+		body: http::Body,
+		decoder: D,
+		decode_buffer: BytesMut,
+		encoder: E,
+		handler: F,
+		finished: bool,
+		_phantom: std::marker::PhantomData<T>,
+	}
+}
+
+pub fn parser<D, E, F, T>(body: http::Body, decoder: D, encoder: E, handler: F) -> http::Body
+where
+	D: Decoder + Send + 'static,
+	D::Error: Send + Into<axum_core::BoxError> + 'static,
+	F: FnMut(D::Item) -> Option<T> + Send + 'static,
+	E: Encoder<T> + Send + 'static,
+	E::Error: Send + Into<axum_core::BoxError> + 'static,
+	T: Send + 'static,
+{
+	http::Body::new(TransformedBody {
+		body,
+		decoder,
+		handler,
+		decode_buffer: BytesMut::new(),
+		encoder,
+		finished: false,
+		_phantom: std::marker::PhantomData,
+	})
+}
+
+impl<D, E, F, T> Body for TransformedBody<D, E, F, T>
+where
+	D: Decoder + Send + 'static,
+	D::Error: Send + Into<axum_core::BoxError> + 'static,
+	E: Encoder<T> + Send + 'static,
+	E::Error: Send + Into<axum_core::BoxError> + 'static,
+	F: FnMut(D::Item) -> Option<T> + Send + 'static,
+{
+	type Data = Bytes;
+	type Error = http::Error;
+
+	fn poll_frame(
+		self: Pin<&mut Self>,
+		cx: &mut Context<'_>,
+	) -> Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
+		let mut this = self.project();
+		// If we're finished and have no more data, we're done
+		if *this.finished {
+			return Poll::Ready(None);
+		}
+
+		let mut encode_buffer = BytesMut::new();
+
+		let mut try_decode = |finished: bool,
+		                      buf: &mut BytesMut,
+		                      decoder: &mut D,
+		                      handler: &mut F,
+		                      encoder: &mut E,
+		                      encode_buf: &mut BytesMut| {
+			loop {
+				let decode = if finished {
+					decoder.decode_eof(buf)
+				} else {
+					decoder.decode(buf)
+				};
+				match decode {
+					Ok(Some(decoded_item)) => {
+						if let Some(transformed_item) = (handler)(decoded_item) {
+							match encoder.encode(transformed_item, encode_buf) {
+								Ok(()) => {},
+								Err(e) => return Err(http::Error::new(e)),
+							}
+						}
+					},
+					Ok(None) => {
+						return Ok(());
+					},
+					Err(e) => {
+						return Err(http::Error::new(e));
+					},
+				}
+			}
+		};
+
+		// Try to decode and encode items from our buffer
+		if let Err(e) = (try_decode)(
+			*this.finished,
+			this.decode_buffer,
+			&mut *this.decoder,
+			this.handler,
+			&mut *this.encoder,
+			&mut encode_buffer,
+		) {
+			return Poll::Ready(Some(Err(e)));
+		}
+
+		// If we have encoded data to send, send it
+		if !encode_buffer.is_empty() {
+			let data = encode_buffer.split_to(encode_buffer.len());
+			return Poll::Ready(Some(Ok(http_body::Frame::data(data.freeze()))));
+		}
+
+		// We need more input data - poll the underlying body
+		let res = ready!(this.body.as_mut().poll_frame(cx));
+		match res {
+			(Some(Ok(frame))) => {
+				if let Some(data) = frame.data_ref() {
```

Comment:
if this is trailers, would they get lost?

Reply:
You are probably the only person I know that can find a bug like this in a 45k line PR :laughing: . Definitely correct, I added a test and a fix